scylla-server-rust/src/main.rs

```diff
@@ -7,44 +7,153 @@ use scylla_server_rust::{
+        // create and spawn the mqtt processor
+        info!("Running processor in MQTT (production) mode");
+        let (recv, eloop) = MqttProcessor::new(
+            tx,
+            std::env::var("PROD_SIREN_HOST_URL").unwrap_or("localhost:1883".to_string()),
+            db.clone(),
+            io,
+        )
+        .await;
+        tokio::spawn(recv.process_mqtt(eloop));
+    }
 
     let app = Router::new()
-        // get all data with the name dataTypeName and runID as specified
+        // DATA ROUTES
         .route(
             "/data/:dataTypeName/:runId",
             get(controllers::data_controller::get_data),
         )
-        // get all datatypes
+        // DATA TYPE ROUTES
         .route("/datatypes", get(data_type_controller::get_all_data_types))
-        // get all drivers
+        // DRIVERS
         .route("/drivers", get(driver_controller::get_all_drivers))
-        // get all locations
+        // LOCATIONS
         .route("/locations", get(location_controller::get_all_locations))
-        // get all nodes
+        // NODES
         .route("/nodes", get(node_controller::get_all_nodes))
-        // runs:
-        // get all runs
+        // RUNS
         .route("/runs", get(run_controller::get_all_runs))
-        // get run with id
         .route("/runs/:id", get(run_controller::get_run_by_id))
-        // get all systems
+        // SYSTEMS
         .route("/systems", get(system_controller::get_all_systems))
+        // for CORS handling
         .layer(
             CorsLayer::new()
                 // allow `GET`
                 .allow_methods([Method::GET])
                 // allow requests from any origin
                 .allow_origin(Any),
         )
-        .with_state(client);
+        // for socketio integration
+        .layer(
+            ServiceBuilder::new()
+                .layer(CorsLayer::permissive())
+                .layer(socket_layer),
+        )
+        .with_state(db.clone());
+
+    let listener = tokio::net::TcpListener::bind("0.0.0.0:8000")
+        .await
+        .expect("Could not bind to 8000!");
+    let axum_token = token.clone();
+    tokio::spawn(async {
+        axum::serve(listener, app)
+            .with_graceful_shutdown(async move {
+                _ = axum_token.cancelled().await;
+            })
+            .await
+            .expect("Failed shutdown init for axum");
+    });
+
+    task_tracker.close();
+
+    info!("Initialization complete, ready...");
+    info!("Use Ctrl+C or SIGINT to exit cleanly!");
 
-    let listener = tokio::net::TcpListener::bind("0.0.0.0:8000").await.unwrap();
-    axum::serve(listener, app).await.unwrap();
+    // listen for ctrl_c, then cancel, close, and await for all tasks in the tracker.  Other tasks cancel vai the default tokio system
+    signal::ctrl_c()
+        .await
+        .expect("Could not read cancellation trigger (ctr+c)");
+    info!("Received exit signal, shutting down!");
+    token.cancel();
```

Comment:
do you have to also cancel the axum_token

Reply:
Nah as it's cloned from the main one and becomes owned by the thread.  It's moved into the thread and the main token is used to cancel.
https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html#:~:text=Struct%20tokio_util%3A%3Async%3A%3ACancellationToken&text=A%20token%20which%20can%20be,resolved%20when%20cancellation%20is%20requested.